File: Solutions/Python/count_complete_tree_nodes.py
"""
    [EASY]
    222. Count Complete Tree Nodes

    Concepts:
    - binary tree
    - depth-first search

    Stats:
        Runtime | 7 ms      [Beats 23.47%]
        Memory  | 23.33 MB  [Beats 10.13%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        count = 0
        
        def dfs(curr):
            nonlocal count
            if not curr:
                return

            count += 1
            dfs(curr.left)
            dfs(curr.right)

        dfs(root)
        return count

File: Solutions/Cpp/count_complete_tree_nodes.cpp
/*
    [EASY]
    222. Count Complete Tree Nodes

    Concepts:
    - binary tree
    - depth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 31.19 MB  [Beats 94.41%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int count = 0;

    void dfs(TreeNode* curr) {
        if (curr == nullptr)
            return;
        
        dfs(curr->left);
        count++;
        dfs(curr->right);
    }

    int countNodes(TreeNode* root) {
        dfs(root);
        return count;
    }
};